File: 백준/Gold/4485. 녹색 옷 입은 애가 젤다지？/녹색 옷 입은 애가 젤다지？.py
# -- METHODS -- #
import heapq
import sys
input = sys.stdin.readline

def dijkstra():
    # 우선순위 큐
    queue = []

    # 시작 지점 초기화, [가중치, 좌표 X, Y)
    heapq.heappush(queue, [plain[0][0], 0, 0])
    cost_array[0][0] = plain[0][0]

    # 좌표축 이동을 위한 direct 변수 초기화
    dx = [-1, 0, 1, 0]
    dy = [0, 1, 0, -1]

    while queue:
        cost, x, y = heapq.heappop(queue)
        if x == n - 1 and y == n - 1:
            break

        # 이미 처리된 좌표인 경우
        if cost_array[x][y] < cost:
            continue

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            # 지도 바깥 범위 예외처리
            if 0 <= nx < n and 0 <= ny < n:
                new_cost = cost + plain[nx][ny]
                # 새로 구한 cost 보다 기존 cost 가 더 작으면 지나감
                if new_cost < cost_array[nx][ny]:
                    cost_array[nx][ny] = new_cost
                    heapq.heappush(queue, [new_cost, nx, ny])

# -- INPUT -- #
itor = 1
INF = int(1e9)
while True:
    n = int(input())
    if n == 0:
        break

    plain = [(list(map(int, input().split()))) for _ in range(n)]  # 평면
    cost_array = [[INF] * n for _ in range(n)]  # 최소 비용 어레이

    dijkstra()
    print(f'Problem {itor}: {cost_array[n - 1][n - 1]}')
    itor += 1

File: 백준/Gold/4485. 녹색 옷 입은 애가 젤다지？/녹색 옷 입은 애가 젤다지？.java

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static int N;
    static int[][] plain, costArr;

    static class Node implements Comparable<Node> {
        int r, c, weight;
        public Node(int r, int c, int weight) {
            this.r = r;
            this.c = c;
            this.weight = weight;
        }
        @Override
        public int compareTo(Node o) {
            return Integer.compare(this.weight, o.weight);
        }
    } // end of class

    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = null;
        StringBuilder sb = new StringBuilder();
        int itor = 0;

        while(true) {

            N = stoi(br.readLine());
            if(N == 0) break;

            plain = new int[N][N];    // 동굴 배열
            costArr = new int[N][N];    // 비용 배열

            for(int i = 0; i < N; i++) {
                st = new StringTokenizer(br.readLine());
                for(int j = 0; j < N; j++) {
                    plain[i][j] = stoi(st.nextToken());
                }
                Arrays.fill(costArr[i], Integer.MAX_VALUE);
            } // end of init

            dijkstra(0, 0);
            sb.append("Problem " + ++itor + ": " + costArr[N - 1][N - 1] + "\n");
        } // end of tc

        System.out.print(sb.toString());
    } // end of main

    static int[] dr = {-1, 0, 1, 0};
    static int[] dc = {0, 1, 0, -1};

    public static void dijkstra(int sr, int sc) {

        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.add(new Node(sr, sc, plain[sr][sc]));
        costArr[sr][sc] = plain[sr][sc];    // 초기 가중치

        while(!pq.isEmpty()) {

            Node current = pq.poll();
            int r = current.r;
            int c = current.c;
            int cost = current.weight;

            if(r == N - 1 && c == N - 1) break;

            for(int d = 0; d < 4; d++) {
                int nr = r + dr[d];
                int nc = c + dc[d];

                if(isPossible(nr, nc)) {
                    if(cost + plain[nr][nc] < costArr[nr][nc]) {
                        costArr[nr][nc] = cost + plain[nr][nc];
                        pq.add(new Node(nr, nc, costArr[nr][nc]));
                    }
                }
            } // end of for
        } // end of while
    } // end of dijkstra

    public static boolean isPossible(int r, int c) {
        return 0 <= r && r < N && 0 <= c && c < N;
    } // end of func

    public static int stoi(String s) {
        return Integer.parseInt(s);
    } // end of stoi
}
